AVX ABI errors when building vector tests

I am trying to build and run tests on commit 61889c8e8926aed2a22000c312baa0a7969cc927 (for https://github.com/dotnet/runtime/pull/50148) and running into `-Wpsabi` errors:

```
$ cat /etc/os-release
NAME=Fedora
VERSION="33 (Workstation Edition)"
ID=fedora
VERSION_ID=33
VERSION_CODENAME=""
PLATFORM_ID="platform:f33"
PRETTY_NAME="Fedora 33 (Workstation Edition)"
ANSI_COLOR="0;38;2;60;110;180"
LOGO=fedora-logo-icon
CPE_NAME="cpe:/o:fedoraproject:fedora:33"
HOME_URL="https://fedoraproject.org/"
DOCUMENTATION_URL="https://docs.fedoraproject.org/en-US/fedora/f33/system-administrators-guide/"
SUPPORT_URL="https://fedoraproject.org/wiki/Communicating_and_getting_help"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_BUGZILLA_PRODUCT="Fedora"
REDHAT_BUGZILLA_PRODUCT_VERSION=33
REDHAT_SUPPORT_PRODUCT="Fedora"
REDHAT_SUPPORT_PRODUCT_VERSION=33
PRIVACY_POLICY_URL="https://fedoraproject.org/wiki/Legal:PrivacyPolicy"
VARIANT="Workstation Edition"
VARIANT_ID=workstation
$ uname -m
x86_64
$ rpm -q clang lldb gcc glibc
clang-11.0.0-2.fc33.x86_64
lldb-11.0.0-1.fc33.x86_64
gcc-10.2.1-9.fc33.x86_64
glibc-2.32-4.fc33.x86_64
$ git rev-parse HEAD~
61889c8e8926aed2a22000c312baa0a7969cc927
$ ./build.sh -subset clr
...
$ ./src/tests/build.sh
...
...
runtime/src/tests/Interop/PInvoke/Generics/GenericsNative.Vector256C.cpp:67:24: error: AVX vector return of type 'Vector256C' (aka '__m256i') without 'avx' enabled changes the ABI [-Werror,-Wpsabi]
    Vector256C value = GetVector256C(e00, e01, e02, e03, e04, e05, e06, e07, e08, e09, e10, e11, e12, e13, e14, e15);
                       ^                                                                                              1 error generated.                                                                                                    
gmake[2]: *** [Interop/PInvoke/Generics/CMakeFiles/GenericsNative.dir/build.make:810: Interop/PInvoke/Generics/CMakeFiles/GenericsNative.dir/GenericsNative.Vector256C.cpp.o] Error 1
gmake[2]: *** Waiting for unfinished jobs....
runtime/src/tests/Interop/PInvoke/Generics/GenericsNative.Vector256L.cpp:43:24: error: AVX vector return of type 'Vector256L' (aka '__m256i') without 'avx' enabled changes the ABI [-Werror,-Wpsabi]
                 Vector256L value = GetVector256L(e00, e01, e02, e03);                                                             
                                    ^                                   
[ 56%] Built target UnmanagedCallersOnlyDll
1 error generated.                                         
gmake[2]: *** [Interop/PInvoke/Generics/CMakeFiles/GenericsNative.dir/build.make:849: Interop/PInvoke/Generics/CMakeFiles/GenericsNative.dir/GenericsNative.Vector256L.cpp.o] Error 1                                         
[ 57%] Built target BoolNative                                                                                        
[ 58%] Built target UIntPtrNative
runtime/src/tests/Interop/PInvoke/Generics/GenericsNative.Vector256B.cpp:99:24: error: AVX vector return of type 'Vector256B' (aka '__m256i') without 'avx' enabled changes the ABI [-Werror,-Wpsabi]
    Vector256B value = GetVector256B(e00, e01, e02, e03, e04, e05, e06, e07, e08, e09, e10, e11, e12, e13, e14, e15, e16, e17, e18, e19, e20, e21, e22, e23, e24, e25, e26, e27, e28, e29, e30, e31);                          
                       ^                                                                      
```

Is this a known issue? Is there a fix/workaround available for this?

--- a/src/tests/Interop/PInvoke/Generics/CMakeLists.txt
+++ b/src/tests/Interop/PInvoke/Generics/CMakeLists.txt
@@ -16,6 +16,10 @@ endif()
 if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
   # The ABI for passing parameters with 32-byte alignment has changed in GCC 4.6
   add_compile_options(-Wno-psabi)
+endif()
+if (NOT MSVC AND NOT CLR_CMAKE_TARGET_ARCH_ARM64 AND NOT CLR_CMAKE_TARGET_ARCH_ARM)
+  # We need -march=native so we can detect if AVX2 is present.
+  # ARM does not like that option too and it make no sense to have this detection there.
   add_compile_options(-march=native)
 endif()
 set(SOURCES

--- a/src/tests/Interop/PInvoke/Generics/CMakeLists.txt
+++ b/src/tests/Interop/PInvoke/Generics/CMakeLists.txt
@@ -1,5 +1,5 @@
-project (GenericsNative) 
-include ("${CLR_INTEROP_TEST_ROOT}/Interop.cmake") 
+project (GenericsNative)
+include ("${CLR_INTEROP_TEST_ROOT}/Interop.cmake")
 if(CLR_CMAKE_TARGET_ARCH_I386)
   add_definitions(-DTARGET_X86)
   add_definitions(-DTARGET_XARCH)
@@ -17,10 +17,15 @@ if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
   # The ABI for passing parameters with 32-byte alignment has changed in GCC 4.6
   add_compile_options(-Wno-psabi)
 endif()
-if (NOT MSVC AND NOT CLR_CMAKE_TARGET_ARCH_ARM64 AND NOT CLR_CMAKE_TARGET_ARCH_ARM)
-  # We need -march=native so we can detect if AVX2 is present.
-  # ARM does not like that option too and it make no sense to have this detection there.
-  add_compile_options(-march=native)
+if (MSVC)
+  # The MSVC require explicitly ask for AVX2 so define would be present.
+  add_compile_options(/arch:AVX2)
+else()
+  if (NOT CLR_CMAKE_TARGET_ARCH_ARM64 AND NOT CLR_CMAKE_TARGET_ARCH_ARM)
+    # We need -march=native so we can detect if AVX2 is present.
+    # ARM does not like that option too and it make no sense to have this detection there.
+    add_compile_options(-march=native)
+  endif()
 endif()
 set(SOURCES
   GenericsNative.IUnknown.cpp
@@ -92,4 +97,4 @@ set(SOURCES
   GenericsNative.VectorU.cpp
 )
 add_library (GenericsNative SHARED ${SOURCES})
-install (TARGETS GenericsNative DESTINATION bin) 
+install (TARGETS GenericsNative DESTINATION bin)

--- a/src/tests/Interop/PInvoke/Generics/CMakeLists.txt
+++ b/src/tests/Interop/PInvoke/Generics/CMakeLists.txt
@@ -17,6 +17,11 @@ if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
   # The ABI for passing parameters with 32-byte alignment has changed in GCC 4.6
   add_compile_options(-Wno-psabi)
 endif()
+if (MSVC)
+    add_compile_options(/arch:AVX2)
+else()
+    add_compile_options(-march=native)
+endif()
 set(SOURCES
   GenericsNative.IUnknown.cpp
   GenericsNative.NullableB.cpp

--- a/src/tests/Interop/PInvoke/Generics/CMakeLists.txt
+++ b/src/tests/Interop/PInvoke/Generics/CMakeLists.txt
@@ -16,11 +16,7 @@ endif()
 if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
   # The ABI for passing parameters with 32-byte alignment has changed in GCC 4.6
   add_compile_options(-Wno-psabi)
-endif()
-if (MSVC)
-    add_compile_options(/arch:AVX2)
-else()
-    add_compile_options(-march=native)
+  add_compile_options(-march=native)
 endif()
 set(SOURCES
   GenericsNative.IUnknown.cpp

--- a/src/tests/Interop/PInvoke/Generics/GenericsNative.Vector256B.cpp
+++ b/src/tests/Interop/PInvoke/Generics/GenericsNative.Vector256B.cpp
@@ -9,6 +9,7 @@
 #if defined(TARGET_XARCH)
     #include <immintrin.h>
 #elif defined(TARGET_ARMARCH)
+    // Intentionally empty
 #else
     #error Unsupported target architecture
 #endif

--- a/src/tests/Interop/PInvoke/Generics/GenericsNative.Vector256B.cpp
+++ b/src/tests/Interop/PInvoke/Generics/GenericsNative.Vector256B.cpp
@@ -8,9 +8,14 @@
 
 #if defined(TARGET_XARCH)
     #include <immintrin.h>
-
-    typedef __m256i Vector256B;
 #elif defined(TARGET_ARMARCH)
+#else
+    #error Unsupported target architecture
+#endif
+
+#if defined(__AVX2__)
+    typedef __m256i Vector256B;
+#else
     typedef struct {
         bool e00;
         bool e01;
@@ -45,8 +50,6 @@
         bool e30;
         bool e31;
     } Vector256B;
-#else
-    #error Unsupported target architecture
 #endif
 
 static Vector256B Vector256BValue = { };

--- a/src/tests/Interop/PInvoke/Generics/GenericsNative.Vector256C.cpp
+++ b/src/tests/Interop/PInvoke/Generics/GenericsNative.Vector256C.cpp
@@ -8,9 +8,14 @@
 
 #if defined(TARGET_XARCH)
     #include <immintrin.h>
-
-    typedef __m256i Vector256C;
 #elif defined(TARGET_ARMARCH)
+#else
+    #error Unsupported target architecture
+#endif
+
+#if defined(__AVX2__)
+    typedef __m256i Vector256C;
+#else
     typedef struct {
         char16_t e00;
         char16_t e01;
@@ -29,8 +34,6 @@
         char16_t e14;
         char16_t e15;
     } Vector256C;
-#else
-    #error Unsupported target architecture
 #endif
 
 static Vector256C Vector256CValue = { };

--- a/src/tests/Interop/PInvoke/Generics/GenericsNative.Vector256C.cpp
+++ b/src/tests/Interop/PInvoke/Generics/GenericsNative.Vector256C.cpp
@@ -9,6 +9,7 @@
 #if defined(TARGET_XARCH)
     #include <immintrin.h>
 #elif defined(TARGET_ARMARCH)
+    // Intentionally empty
 #else
     #error Unsupported target architecture
 #endif

--- a/src/tests/Interop/PInvoke/Generics/GenericsNative.Vector256D.cpp
+++ b/src/tests/Interop/PInvoke/Generics/GenericsNative.Vector256D.cpp
@@ -9,6 +9,7 @@
 #if defined(TARGET_XARCH)
     #include <immintrin.h>
 #elif defined(TARGET_ARMARCH)
+    // Intentionally empty
 #else
     #error Unsupported target architecture
 #endif

--- a/src/tests/Interop/PInvoke/Generics/GenericsNative.Vector256D.cpp
+++ b/src/tests/Interop/PInvoke/Generics/GenericsNative.Vector256D.cpp
@@ -8,17 +8,20 @@
 
 #if defined(TARGET_XARCH)
     #include <immintrin.h>
-
-    typedef __m256d Vector256D;
 #elif defined(TARGET_ARMARCH)
+#else
+    #error Unsupported target architecture
+#endif
+
+#if defined(__AVX2__)
+    typedef __m256 Vector256D;
+#else
     typedef struct {
         double e00;
         double e01;
         double e02;
         double e03;
     } Vector256D;
-#else
-    #error Unsupported target architecture
 #endif
 
 static Vector256D Vector256DValue = { };

--- a/src/tests/Interop/PInvoke/Generics/GenericsNative.Vector256F.cpp
+++ b/src/tests/Interop/PInvoke/Generics/GenericsNative.Vector256F.cpp
@@ -9,6 +9,7 @@
 #if defined(TARGET_XARCH)
     #include <immintrin.h>
 #elif defined(TARGET_ARMARCH)
+    // Intentionally empty
 #else
     #error Unsupported target architecture
 #endif

--- a/src/tests/Interop/PInvoke/Generics/GenericsNative.Vector256F.cpp
+++ b/src/tests/Interop/PInvoke/Generics/GenericsNative.Vector256F.cpp
@@ -8,9 +8,14 @@
 
 #if defined(TARGET_XARCH)
     #include <immintrin.h>
-
-    typedef __m256 Vector256F;
 #elif defined(TARGET_ARMARCH)
+#else
+    #error Unsupported target architecture
+#endif
+
+#if defined(__AVX2__)
+    typedef __m256 Vector256F;
+#else
     typedef struct {
         float e00;
         float e01;
@@ -21,8 +26,6 @@
         float e06;
         float e07;
     } Vector256F;
-#else
-    #error Unsupported target architecture
 #endif
 
 static Vector256F Vector256FValue = { };

--- a/src/tests/Interop/PInvoke/Generics/GenericsNative.Vector256L.cpp
+++ b/src/tests/Interop/PInvoke/Generics/GenericsNative.Vector256L.cpp
@@ -9,6 +9,7 @@
 #if defined(TARGET_XARCH)
     #include <immintrin.h>
 #elif defined(TARGET_ARMARCH)
+    // Intentionally empty
 #else
     #error Unsupported target architecture
 #endif

--- a/src/tests/Interop/PInvoke/Generics/GenericsNative.Vector256L.cpp
+++ b/src/tests/Interop/PInvoke/Generics/GenericsNative.Vector256L.cpp
@@ -8,17 +8,20 @@
 
 #if defined(TARGET_XARCH)
     #include <immintrin.h>
-
-    typedef __m256i Vector256L;
 #elif defined(TARGET_ARMARCH)
+#else
+    #error Unsupported target architecture
+#endif
+
+#if defined(__AVX2__)
+    typedef __m256i Vector256L;
+#else
     typedef struct {
         int64_t e00;
         int64_t e01;
         int64_t e02;
         int64_t e03;
     } Vector256L;
-#else
-    #error Unsupported target architecture
 #endif
 
 static Vector256L Vector256LValue = { };

--- a/src/tests/Interop/PInvoke/Generics/GenericsNative.Vector256U.cpp
+++ b/src/tests/Interop/PInvoke/Generics/GenericsNative.Vector256U.cpp
@@ -9,6 +9,7 @@
 #if defined(TARGET_XARCH)
     #include <immintrin.h>
 #elif defined(TARGET_ARMARCH)
+    // Intentionally empty
 #else
     #error Unsupported target architecture
 #endif

--- a/src/tests/Interop/PInvoke/Generics/GenericsNative.Vector256U.cpp
+++ b/src/tests/Interop/PInvoke/Generics/GenericsNative.Vector256U.cpp
@@ -8,9 +8,14 @@
 
 #if defined(TARGET_XARCH)
     #include <immintrin.h>
-
-    typedef __m256i Vector256U;
 #elif defined(TARGET_ARMARCH)
+#else
+    #error Unsupported target architecture
+#endif
+
+#if defined(__AVX2__)
+    typedef __m256i Vector256U;
+#else
     typedef struct {
         uint32_t e00;
         uint32_t e01;
@@ -21,8 +26,6 @@
         uint32_t e06;
         uint32_t e07;
     } Vector256U;
-#else
-    #error Unsupported target architecture
 #endif
 
 static Vector256U Vector256UValue = { };

